### DISCUSSION TITLE
Streaming JSON data from an Edge function

### DISCUSSION THREAD

**Discussion by jschuur** [vercel/next.js#48380] [Help]
### Summary

I'm trying to stream JSON data in a long running Edge function (via a `ReadableStream`), but I'm running into issues because sometimes multiple queued chunks are sent out together (which breaks the JSON stringified encoding) and I don't see a way to force an encoded chunk to be sent out.

There is also an 'object mode' for streams but this seems to apply to `Readable` instances and not the `ReadableStream` that I'm using.

### Additional information

I have an edge function that looks like this:

```js
const feedCandidates = ['rss.xml', 'feed.xml', 'atom.xml', 'feed.rss', 'feed.atom'];

export const config = {
  runtime: 'edge',
};

async function validateUrl(url: string): Promise<boolean> {
  const result = await fetch(url, { method: 'HEAD', cache: 'no-store' });

  return result.status === 200 || false;
}

export async function POST(request: Request) {
  const encoder = new TextEncoder();

  const stream = new ReadableStream({
    async start(controller) {
      for (const candidate of feedCandidates) {
        const url = `https://turbo.build/${candidate}`;
        controller.enqueue(encoder.encode(`Looking up ${url}...`));

        const result = (await validateUrl(url as string)) ? 'OK' : 'FAIL';
        controller.enqueue(encoder.encode(`Looking up ${url}... (${result})`));
      }

      controller.close();
    },
  });

  return new Response(stream);
}
```

Combined with a custom hook that makes the post request and a component to show the data as chunks come in (using `getReader` and `TextDecoder`, this generates an output like this:

```
Looking up https://turbo.build/rss.xml...
Looking up https://turbo.build/rss.xml... (FAIL)Looking up https://turbo.build/feed.xml...
Looking up https://turbo.build/feed.xml... (OK)Looking up https://turbo.build/atom.xml...
Looking up https://turbo.build/atom.xml... (FAIL)Looking up https://turbo.build/feed.rss...
Looking up https://turbo.build/feed.rss... (FAIL)Looking up https://turbo.build/feed.atom...
Looking up https://turbo.build/feed.atom... (FAIL)
```

Note that in the output each newline is a new chunk received on the client side.

So the first one was `Looking up https://turbo.build/rss.xml...` and the second one was `Looking up https://turbo.build/rss.xml... (FAIL)Looking up https://turbo.build/feed.xml...`. Someone my first outgoing chunk went out right away, but all subsequent ones were paired up with the next one.

Client side console logs also confirm that the chunks arrived merge together.

You can imagine if I'd JSON stringified these chunks that the concatenated ones wouldn't have parsed.

I'm not sure if [`objectMode`](https://nodejs.org/api/stream.html#implementing-a-readable-stream) would help here, but I also don't understand enough of streams to rewrite this to use `stream.Readable`.

### Example

https://github.com/jschuur/next-edge-streaming-bug

**Comment by jschuur** [comment]
Happens with both Next.js 13.3 app dir API routes and pages ones.

If I hardcode how I assign `result` with a static string instead of the async function call response, it sends everything in a single chunk, so it looked like timing was a factor, but if I add a longer delay before the second `enqueue` it still kept combining two chunks.

Its seems to be ignoring `{ highWaterMark: 1 }` (although technically that doesn't mean 'send after 1 chunk', it just means 'apply backpressure', right?). Unless this is just the setting for the stream that feeds into something the actual HTTP stream generated and I need to find a way to control the same setting there.

**Comment by jschuur** [comment]
This seems to be relevant https://vercel.com/docs/concepts/functions/edge-functions/streaming#caveats.

It suggests ways to handle chunks that may or may not be JSON parseable yet. But that doesn't help me when too many chunks are buffered in the first place and I want to force them from going out one by one.

I'm starting to see that a `ReadableStream` doesn't have the concept of being forced to send what's in its buffer. So at this point, the question is can I do this at any other point in the streaming process?

Otherwise, I can't iterate through an array and provide individual before and after messages for an operation on an element that are seen in realtime. I would have to use websockets and couldn't do that within the API route options of Next.js.

**Comment by ajayvignesh01** [comment]
This is just how ReadableStream works. You don't have control over how the chunks are set. It's dynamically adjusted based on network conditions (ex: how far away servers are or how fast the connection is)

**Comment by marclelamy** [comment]
hey @jschuur did you find a work around?

**Comment by ericzakariasson** [comment]
My best solution to this at this point has been to suffix with a newline 
```ts
const encoder = new TextEncoder();
const encoded = encoder.encode(`${JSON.stringify(obj)}\n`)
controller.enqueue(encoded)
```

and in the consumer you'll have to expect partial chunks, something like this:

```ts
const decoder = new TextDecoder();

let partial = "";

while (true) {
  const { done, value } = await reader.read();
  if (done) {
    break;
  }

  const decoded = decoder.decode(value, { stream: true });
  const lines = decoded.split(SKYFFEL_LINE_SEPARATOR).filter(Boolean);


  for (const line of lines) {
    try {
      const json = JSON.parse(partial);
      yield json
    } catch (error) {
      partial += line;
    }
  }
}
```

Not the prettiest, but it works
